File: 1127-last-stone-weight/1127-last-stone-weight.py
import heapq
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        if len(stones)==0:
            return 0
        if len(stones)==1:
            return stones[0]
        
        heap=[-x for x in stones]
        heapq.heapify(heap)

        while len(heap)>1:
            y=heapq.heappop(heap)
            x=heapq.heappop(heap)
            if x==y:
                continue
            else:
                heapq.heappush(heap, y-x)
            
        val= 0 if len(heap)==0 else -1 * heap[0]
        
        return val

File: 1127-last-stone-weight/1127-last-stone-weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pQueue=new PriorityQueue<>((a,b)->b-a);
        for(int stone:stones){
            pQueue.offer(stone);
        }
        while(pQueue.size()>1){
            int y=pQueue.poll();
            int x=pQueue.poll();
            if(x==y){
                continue;
            }
            pQueue.offer(y-x);
        }
        return pQueue.size()==1? pQueue.poll() : 0;
    }
}